pandas/core/indexes/base.py

```diff
@@ -5409,6 +5395,19 @@ def _find_common_type_compat(self, target) -> DtypeObj:
             return IntervalDtype(np.float64, closed=self.closed)
 
         target_dtype, _ = infer_dtype_from(target, pandas_dtype=True)
+
+        # special case: if one dtype is uint64 and the other a signed int, return object
+        # See https://github.com/pandas-dev/pandas/issues/26778 for discussion
+        # Now it's:
+        # * float | [u]int -> float
+        # * uint64 | signed int  -> object
+        # We may change union(float | [u]int) to go to object.
+        if self.dtype == "uint64" or target_dtype == "uint64":
+            if is_signed_integer_dtype(self.dtype) or is_signed_integer_dtype(
+                target_dtype
+            ):
```

Comment:
in principle we could check `signed_one.min()` and if its non-negative could avoid the object-dtype cast.  not for this PR.